File: Httbar/myscripts/Httbar_clcheng/Tool/ScaleAllShapes.py
#!/usr/bin/env python
from argparse import ArgumentParser
import numpy as np
parser = ArgumentParser()
parser.add_argument('input',nargs='+', help = "Input templates for scaling")
parser.add_argument(
    '--factor',default = 1000./35.9,type = float,help="""Factor by which all shapes are scaled (sgn, int, bkg)""")  
args = parser.parse_args()

import ROOT
import shutil
from pdb import set_trace

factor = args.factor

for files in args.input:
 print "Loading file {0} Scaling factor: {1}".format(files,args.factor)
 f = files.replace(".root","_all_scaled.root")
 shutil.copyfile(files, f)
 f = ROOT.TFile(f,"UPDATE")
 for category in [i.GetName() for i in f.GetListOfKeys()]:
  tdir = f.Get(category)
  tdir.cd()
  shapes = [i.GetName() for i in tdir.GetListOfKeys()]
  n = len(shapes)
  counter = 0
  print "Doing category: {0} # of shapes {1}".format(category,n)
  for shape in shapes:
   h = tdir.Get(shape)
   h.Scale(factor)
   h.Write(shape,ROOT.TObject.kOverwrite)
   counter += 1
   print "Doing shapes: {0}/{1}".format(counter,n)
 f.Close()   

File: Httbar/myscripts/Httbar_clcheng/Tool/ScaleAllShapes.c
#include "TString.h"
#include "TObjArray.h"
#include "TObjString.h"
#include <TSystem.h>
#include <iostream>
#include "TFile.h"
#include "TDirectoryFile.h"
#include "TH1F.h"
#include "TObject.h"
#include <signal.h>

//Usage: ScaleAllHist("template_sgn.root,template_int.root",100./35.9)

int ScaleAllHist(TString input, Double_t factor){
 TObjArray *files = input.Tokenize(",");
 for (Int_t i = 0; i < files->GetEntries(); i++) {
   TString file = ((TObjString*)files->At(i))->String();
   TString nfile1 = file;
   nfile1.ReplaceAll(".root","");
   TString nfile ;
   nfile.Form("%s_scaled_by_%.2f.root",nfile1.Data(),factor);
   gSystem->CopyFile(file,nfile,-1);
   TFile* f = new TFile(nfile,"UPDATE");
   std::cout<<"Loading file: "<<file<<" Scaling factor: "<<factor<<"\n";
   std::cout<<"Update file: "<<nfile<<" Scaling factor: "<<factor<<"\n";
   TList* l = f->GetListOfKeys();
   for (Int_t i = 0; i < l->GetEntries(); i++) {
     TString category = l->At(i)->GetName();
     TDirectoryFile* tdir = (TDirectoryFile*) f->Get(category);
     tdir->cd();
     TList* shapes = tdir->GetListOfKeys();
     std::cout<<"Doing category: "<<category<<"\n";
     Int_t n = shapes->GetEntries();
     for (Int_t j = 0; j < n; j++) {
       TString shape = shapes->At(j)->GetName();
       TH1F * h = (TH1F *)tdir->Get(shape);
       h->Scale(factor);
       h->Write(shape,TObject::kWriteDelete);
     } 
   }
   f->Close();
   delete f;
 }
 return 0;
}
